### PR TITLE
2d primitive tests

### PR DESCRIPTION
**Pull request by abairmj**
Hey bro can you check my work?  I tried to clean it up and test colors on the lines.  Wanted to see what it looks like on the xbox hardware.  

Called you for some git advice on whether to squash or not.  Looks like that made your material_tests changes show up in this PR?  

Also wanted to see if you knew a way for me to test if 
0x3 for NV04_SOLID_LINE_COLOR_FORMAT is actually NV097_SET_SURFACE_FORMAT_COLOR_LE_R5G6B5?

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -16,11 +16,13 @@
 
 #include "test_driver.h"
 #include "test_host.h"
+
```
Comment (abaire) [inline main.cpp:19]: Nit: remove (I think `clang-format` might do this for you, you may need to run it manually since I'm not sure if the hook re-checks all the files on an amend).

File: pbkit_ext.h [hunk 2]
```diff
@@ -14,6 +14,11 @@
 // "Beta4" class for blending operations (see xf86-video-nouveau).
 #define GR_CLASS_72 0x72
 
+// Enumerated DMA channel destinations
+#define DMA_CHANNEL_FOR_PIXEL_RENDERER_9 9
```
Comment (abaire) [inline pbkit_ext.h:18]: Let's drop the numbers from these constants, they're not actually relevant for users. I also think we can be slightly more terse and drop the `FOR` part.

e.g., `DMA_CHANNEL_PIXEL_RENDERER`

File: tests/two_d_primitive_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
+#include "two_d_primitive_tests.h"
+
+#include <SDL_image.h>
```
Comment (abaire) [inline tests/two_d_primitive_tests.cpp:3]: I think this code no longer uses SDL.

File: tests/two_d_primitive_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+#include <pbkit/pbkit.h>
+
+#include "../nxdk_ext.h"
+#include "../pbkit_ext.h"
+#include "../test_host.h"
+#include "debug_output.h"
+#include "vertex_buffer.h"
+
+// From pbkit.c
+#define MAXRAM 0x03FFAFFF
```
Comment (abaire) [inline tests/two_d_primitive_tests.cpp:13]: And this test doesn't set up DMA so it doesn't need this constant. 

We might also want to move the constant to pbkit_ext.h if you're interested, but that should be a separate PR.

File: tests/two_d_primitive_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+// See pb_init in pbkit.c, where the channel contexts are set up.
+// SUBCH_4 == GR_CLASS_62, which contains the NV10 2D surface commands.
+#define SUBCH_CLASS_62 SUBCH_4
```
Comment (abaire) [inline tests/two_d_primitive_tests.cpp:17]: Unused

File: tests/two_d_primitive_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    {0xFF0000, 400, 200, 400, 400},
+    {0x00FF00, 400, 200, 400, 400},
+    {0x0000FF, 400, 200, 400, 400},
+    {0xFFFFFF, 0, 0, 799, 599},
+    {0xFFFFFF, 799, 599, 0, 0},
+    {0xFFFFFF, 400, 0, 400, 599},
+    {0xFFFFFF, 400, 300, 401, 300},
+    {0xFFFFFF, 799, 599, 0, 0},
+    {0xFFFFFF, 999, 300, 0, 300},
+};
```
Comment (abaire) [inline tests/two_d_primitive_tests.cpp:35]: Might also be interesting to look at negative numbers?

File: tests/two_d_primitive_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+  pb_create_gr_ctx(channel++, NV04_SOLID_LINE, &solid_lin_ctx_);
+  pb_bind_channel(&solid_lin_ctx_);
+  pb_bind_subchannel(SUBCH_CLASS_5C, &solid_lin_ctx_);
+
+  pb_create_gr_ctx(channel++, NV04_CONTEXT_SURFACES_2D, &surface_destination_ctx_);
+  pb_bind_channel(&surface_destination_ctx_);
+  pb_bind_subchannel(SUBCH_CLASS_42, &surface_destination_ctx_);
+
+  // set up to not have shader
```
Comment (abaire) [inline tests/two_d_primitive_tests.cpp:60]: In this case we're not actually using the 3d hardware at all, so maybe drop these comments (we still should probably clear out the shader program since it does some changes to the combiner that may or may not be important)

File: tests/two_d_primitive_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+  // no vertex shader (transform shader)..
+  // no fragment or pixel shader (color from triangle)..
+  host_.SetShaderProgram(nullptr);
+}
+
+void TwoDPrimitiveTests::Test(const TwoDPrimitiveTest& test) {
+  host_.PrepareDraw(0xFF440011); //beta + RRGGBB
```
Comment (abaire) [inline tests/two_d_primitive_tests.cpp:67]: This is actually `AARRGGBB`, beta is a blend factor whereas alpha is the actual transparency value.

File: tests/two_d_primitive_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+  auto p = pb_begin();
+
+  p = pb_push1_to(SUBCH_CLASS_42, p, NV04_CONTEXT_SURFACES_2D_SET_DMA_IMAGE_DST, DMA_CHANNEL_FOR_PIXEL_RENDERER_9);
+
+#define NV042_SET_PITCH 0x304
```
Comment (abaire) [inline tests/two_d_primitive_tests.cpp:73]: Let's move the constants into nxdk_ext.h if they're for sure correct and needed.

File: tests/two_d_primitive_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+  host_.FinishDraw();
+  // host_.FinishDrawAndSave(output_dir_.c_str(), name.c_str());
+}
+
+std::string TwoDPrimitiveTests::MakeTestName(const TwoDPrimitiveTest& test) {
+  char buf[256] = {0};
+  snprintf(buf, 255, "2D Line - %08X - Sx%lu Sy%lu - Ex%lu Ey%lu",
+           test.two_d_object_color, test.two_d_start_x, test.two_d_start_y, test.two_d_end_x, test.two_d_end_y);
+  return buf;
+}
```
Comment (abaire) [inline tests/two_d_primitive_tests.cpp:106]: Nit: newline at end of file (`clang-format` might fix this)

File: tests/two_d_primitive_tests.cpp [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+  pb_draw_text_screen();
+  //  std::string name = MakeTestName(test);
+
+  host_.FinishDraw();
+  // host_.FinishDrawAndSave(output_dir_.c_str(), name.c_str());
+}
+
+std::string TwoDPrimitiveTests::MakeTestName(const TwoDPrimitiveTest& test) {
+  char buf[256] = {0};
+  snprintf(buf, 255, "2D Line - %08X - Sx%lu Sy%lu - Ex%lu Ey%lu",
```
Comment (abaire) [inline tests/two_d_primitive_tests.cpp:103]: This might end up generating a filename that is too long and will blow up when `FinishDrawAndSave` is called. The current convention is also to not have any spaces in the filename, so I'd do underscores and be really terse. 

As an aside, once #8 is done we can probably revisit naming and reclaim some of the chars used for test suite names.

File: tests/two_d_primitive_tests.h [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+#include <pbkit/pbkit.h>
+
+#include "test_suite.h"
+
+class TestHost;
+
+class TwoDPrimitiveTests : public TestSuite {
+ public:
+  struct TwoDPrimitiveTest {
+    uint32_t two_d_object_color;
```
Comment (abaire) [inline tests/two_d_primitive_tests.h:13]: `two_d` is redundant for all of these, and for this one I'd even drop "object", just `color`

File: tests/two_d_primitive_tests.h [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    uint32_t two_d_start_x;
+    uint32_t two_d_start_y;
+    uint32_t two_d_end_x;
+    uint32_t two_d_end_y;
+  };
+
+ public:
+  TwoDPrimitiveTests(TestHost& host, std::string output_dir);
+
+  std::string Name() override { return "2D Primitive"; }
```
Comment (abaire) [inline tests/two_d_primitive_tests.h:23]: If you plan on doing more than just lines in this suite I think this name is fine, but if you want to test other primitives elsewhere maybe just "2D Lines"?

Comment (abairmj) [reply tests/two_d_primitive_tests.h:23]: Originally I thought I would be able to do rectangle's and others but I never saw those exposed in the header files.  I think for now I'll convert to LINES (although I'm not sure yet if we're creating lines in hardware at all or just LIN's ;)

File: pbkit_ext.h (hunk 2 above)
Comment (abairmj) [reply pbkit_ext.h:18]: I was worried if I didn't include the 9 if you were searching for a DMA by one of the known channel values you'd miss the hit.  But I can whack them if you'd like.  I also wasn't sure if this belonged in pbkit_ext or the nxdk_ext.

File: tests/two_d_primitive_tests.cpp (hunk 3 above)
Comment (abairmj) [reply tests/two_d_primitive_tests.cpp:3]: good catch, intellij had the includes collapsed on. me.  removed.

Comment (abairmj) [reply tests/two_d_primitive_tests.cpp:3]: I'm thinking maybe I don't need vertexbuffer.h or debugoutput.h either?  @abaire am I good to remove those as well?

File: tests/two_d_primitive_tests.cpp (hunk 5 above)
Comment (abairmj) [reply tests/two_d_primitive_tests.cpp:17]: done

File: tests/two_d_primitive_tests.cpp (hunk 6 above)
Comment (abairmj) [reply tests/two_d_primitive_tests.cpp:35]: I was thinking the same but then I didn't know if it would do anything since the coords were uint32_t's?.

File: tests/two_d_primitive_tests.cpp (hunk 7 above)
Comment (abairmj) [reply tests/two_d_primitive_tests.cpp:60]: done, left them in for learning in case I needed to remember the different shader types and what we might use them from.  Did you learn those from envytools docs?

File: tests/two_d_primitive_tests.cpp (hunk 8 above)
Comment (abairmj) [reply tests/two_d_primitive_tests.cpp:67]: fixed

File: pbkit_ext.h (hunk 2 above)
Comment (abaire) [reply pbkit_ext.h:18]: pbkit_ext is the right home since this is tied directly to the pbkit implementation (I think it could decide to use any DMA channel for this)

In practice searching for the 9 is likely to have so many results that it'd be noise anyway, and if pbkit gets refactored in the future such that DMA_CHANNEL_PIXEL_RENDERER becomes `15` you'd end up needing to rename things scattered across the codebase that don't actually care what the value is, just what the logical use case is

File: tests/two_d_primitive_tests.cpp (hunk 3 above)
Comment (abaire) [reply tests/two_d_primitive_tests.cpp:3]: Yeah, please remove anything that's not actually used. I think CLion does a reasonably good job at greying out unused headers, so you can just delete them.

File: tests/two_d_primitive_tests.cpp (hunk 6 above)
Comment (abaire) [reply tests/two_d_primitive_tests.cpp:35]: Yeah, you'd have to switch to signed ints or just use twos complement values.

File: pbkit_ext.h (hunk 2 above)
Comment (abairmj) [reply pbkit_ext.h:18]: done

File: tests/two_d_primitive_tests.cpp (hunk 7 above)
Comment (abaire) [reply tests/two_d_primitive_tests.cpp:60]: The pipeline is pretty much standard DirectX or OpenGL, there's not a whole lot of nv2a stuff at that level other than restrictions on the shader model (which is documented by Nvidia and MS outside of envytools)

File: tests/two_d_primitive_tests.cpp (hunk 3 above)
Comment (abairmj) [reply tests/two_d_primitive_tests.cpp:3]: done

File: tests/two_d_primitive_tests.cpp (hunk 9 above)
Comment (abairmj) [reply tests/two_d_primitive_tests.cpp:73]: done

File: main.cpp (hunk 1 above)
Comment (abairmj) [reply main.cpp:19]: done

File: tests/two_d_primitive_tests.cpp (hunk 4 above)
Comment (abairmj) [reply tests/two_d_primitive_tests.cpp:13]: done

File: tests/two_d_primitive_tests.cpp (hunk 10 above)
Comment (abairmj) [reply tests/two_d_primitive_tests.cpp:106]: done

File: tests/two_d_primitive_tests.cpp (hunk 11 above)
Comment (abairmj) [reply tests/two_d_primitive_tests.cpp:103]: done

File: tests/two_d_primitive_tests.h (hunk 12 above)
Comment (abairmj) [reply tests/two_d_primitive_tests.h:13]: done

File: tests/two_d_primitive_tests.h (hunk 13 above)
Comment (abairmj) [reply tests/two_d_primitive_tests.h:23]: done

File: tests/two_d_line_tests.cpp [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    {0x0000FF, 400, 200, 400, 400},
+    {0xFFFFFF, 0, 0, 799, 599},
+    {0xFFFFFF, 799, 599, 0, 0},
+    {0xFFFFFF, 400, 0, 400, 599},
+    {0xFFFFFF, 400, 300, 401, 300},
+    {0xFFFFFF, 799, 599, 0, 0},
+    {0xFFFFFF, 999, 300, 0, 300},
+};
+// clang-format on
+TwoDLineTests::TwoDLineTests(TestHost& host, std::string output_dir) : TestSuite(host, std::move(output_dir)) {
```
Comment (abaire) [inline tests/two_d_line_tests.cpp:27]: Nit add a blank line

File: tests/two_d_line_tests.cpp [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+// clang-format off
+static constexpr TwoDLineTests::TestCase kTests[] = {
+    {0x000000, 400, 200, 400, 400},
+    {0xFF0000, 400, 200, 400, 400},
+    {0x00FF00, 400, 200, 400, 400},
+    {0x0000FF, 400, 200, 400, 400},
+    {0xFFFFFF, 0, 0, 799, 599},
+    {0xFFFFFF, 799, 599, 0, 0},
+    {0xFFFFFF, 400, 0, 400, 599},
+    {0xFFFFFF, 400, 300, 401, 300},
```
Comment (abaire) [inline tests/two_d_line_tests.cpp:23]: Might also be interesting to test a point (start == end)

**Review by abaire — APPROVED** [review]

**Review by abaire — APPROVED** [review]
